web/src/osm2streets-svelte/Layer.svelte

```diff
@@ -0,0 +1,158 @@
+
+  // Input
+  export let source: string;
+  export let show = true;
+  export let gj: GeoJSON | undefined;
+  // TODO LayerSpecification doesn't work
+  export let layerStyle: any;
+  // Make objects hoverable and clickable. The caller should do something with
+  // ["feature-state", "hover"] and ["feature-state", "clicked"]
+  export let interactive = false;
+
+  // Output
+  export let hoveredFeature: Feature | null = null;
+  export let clickedFeature: Feature | null = null;
+
+  // An opaque ID assigned by this component
+  let hoverId: number | undefined;
+  let clickedId: number | undefined;
+
+  let layer = `${source}-layer`;
+
+  onMount(() => {
+    fixIDs();
+    $map!.addSource(source, {
+      type: "geojson",
+      data: gj ?? emptyGeojson(),
+    });
+    $map!.addLayer(
+      {
+        id: layer,
+        source,
+        ...layerStyle,
+      },
+      getLayerZorder(layer)
+    );
+    // We may need to hide initially
+    if (!show) {
+      $map!.setLayoutProperty(layer, "visibility", "none");
+    }
+
+    if (interactive) {
+      // Configure hovering
+      $map!.on("mousemove", layer, onMouseMove);
+      $map!.on("mouseleave", layer, onMouseLeave);
+      // Configure clicking
+      $map!.on("click", onClick);
+    }
+  });
+
+  onDestroy(() => {
+    if (interactive) {
+      $map!.off("mousemove", onMouseMove);
+      $map!.off("mouseleave", onMouseLeave);
+      $map!.off("click", onClick);
+      unhover();
+    }
+
+    if ($map!.getLayer(layer)) {
+      $map!.removeLayer(layer);
+    }
+    $map!.removeSource(source);
+  });
+
+  // Make sure every feature has a numeric ID. But don't use generateId,
+  // because we need to hold onto the full GeoJSON we give to MapLibre,
+  // because there's not a way to get it back out later.
+  function fixIDs() {
+    // If we've only been passed in one feature, don't bother with IDs
+    if (gj && "features" in gj) {
+      for (let [idx, f] of gj.features.entries()) {
+        // 0 is problematic
+        f.id = idx + 1;
+      }
+    }
+    // DON'T do 'gj = gj', to avoid the reactive statement below running spuriously
+  }
+
+  $: {
+    let sourceObj = $map.getSource(source);
+    if (sourceObj) {
+      console.log(`GeoJSON data for ${source} changed, updating`);
+      fixIDs();
+      (sourceObj as GeoJSONSource).setData(gj ?? emptyGeojson());
+
+      hoveredFeature = null;
+      hoverId = undefined;
+      clickedFeature = null;
+      clickedId = undefined;
+    }
+  }
+
+  // Show/hide
+  $: {
+    if ($map!.getLayer(layer)) {
+      $map!.setLayoutProperty(layer, "visibility", show ? "visible" : "none");
+    }
+  }
+
+  function onMouseMove(e: MapLayerMouseEvent) {
+    if (e.features.length > 0 && hoverId != e.features[0].id) {
+      unhover();
+      // generateId means this'll be a number
+      hoverId = e.features[0].id as number;
+      $map!.setFeatureState({ source, id: hoverId }, { hover: true });
+
+      // Per
```

Comment:
This kind of thing is a very annoying aspect of MapLibre's API, but we can work around it cleanly here